keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,431 @@
+   details: a) how determining node/instance health can best be done for
+   Kubernetes clusters b) how Kube-proxy will do it once the changes proposed in
+   this KEP are merged c) what some pitfalls with other methods might be.
+
+### Test Plan
+
+#### Prerequisite testing updates
+
+#### Unit tests
+
+Update the Kube-proxy unit tests to include the `healthz` answer for the
+healthcheck server test suite.
+
+- `k8s.io/kubernetes/pkg/proxy/healthcheck`: `09/Feb/2023` - `68.8%`
+
+#### Integration tests
+
+This feature is not readily integration tested, so we will use unit and E2E.
+
+#### e2e tests
+
+- Add E2E tests for connectivity to services on terminating nodes and
+  validate graceful termination of TCP connections.
+
+### Graduation Criteria
+
+#### Alpha
+
+- E2E tests coded before any feature implementation is made which highlights the
+  existing problem.
+- Decision on final field on Node object to be used as an indicator of "node is
+  terminating/deleting"
+- Feature implemented behind a feature flag.
+- Document written at https://kubernetes.io/docs/concepts/
+
+#### Beta
+
+- No issues reported.
+
+#### GA
+
+- No issues reported during two releases.
+
+### Upgrade / Downgrade Strategy
+
+Any upgrade to a version enabling the feature, succeeded by a downgrade to a
+version disabling it, is not expected to be impact ingress in any way, given
+that Kube-proxy is healthy on all cluster nodes. Should any Kube-proxy not be
+healthy: then ingress for `eTP:Cluster` services won't be using that node as a
+nexthop for ingress traffic. This would have been the case in the preceding
+version
+
+### Version Skew Strategy
+
+This doesn't touch load balancer / HC API, so even though an old Kube-proxy
+might talk to a newer control plane, there's no real concern.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `KubeProxyDrainingTerminatingNodes`
+  - Components depending on the feature gate: Kube-proxy
+
+###### Does enabling the feature change any default behavior?
+
+Yes. For `eTP:Cluster` services: Kube-proxy currently doesn't include any logic
+about terminating / deleting nodes when determining if it's healthy. This will
+be the case going forward, whereby the addition of the taint
+`ToBeDeletedByClusterAutoscaler` will cause Kube-proxy to fail its `healthz`.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, by resetting the feature gate back.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Behavior will be restored back immediately.
+
+###### Are there any tests for feature enablement/disablement?
+
+Not needed since the enablement/disablement is triggered by changing a in-memory boolean
+variable.
```

Comment:
nit: what matters is that the feature is purely in-memory (i.e. no persistent data is storage based on it being enabled/disabled), not the fact that it's enablement/disablement is.

Not blocking, but let's maybe change to:
`
Not needed, since the feature is purely in-memory thing with no consequences for any persistent data.
`

Reply:
Oh right, the question makes much more sense now :)